File: 04-stack/00150-Evaluate_Reverse_Polish_Notation/00150-evaluate_reverse_polish_notation.cpp
// 150: Evaluate Reverse Polish Notation
// https://leetcode.com/problems/evaluate-reverse-polish-notation/


#include <iostream>
#include <vector>
#include <stack>

using namespace std;


class Solution {
public:
    // SOLUTION
    int evalRPN(vector<string>& tokens) {
        stack<int> result;

        for (auto s : tokens) {
            if (s.size()>1 || isdigit(s[0])) {
                result.push(stoi(s));
            } else {
                auto x2 = result.top(); result.pop();
                auto x1 = result.top(); result.pop();
                
                switch (s[0]) {
                    case '+': x1+=x2; break;
                    case '-': x1-=x2; break;
                    case '*': x1*=x2; break;
                    case '/': x1/=x2; break;
                }

                result.push(x1);
            }
        }

        return result.top();
    }
};



int main() {
    Solution o;

    // INPUT
    vector<string> tokens = {"2","1","+","3","*"};

    // OUTPUT
    auto result = o.evalRPN(tokens);
    cout<<result<<endl;

    return 0;
}

File: 04-stack/00150-Evaluate_Reverse_Polish_Notation/00150-evaluate_reverse_polish_notation.py
# 150: Evaluate Reverse Polish Notation
# https://leetcode.com/problems/evaluate-reverse-polish-notation/


class Solution:
    # SOLUTION
    def evalRPN(self, tokens: list[str]) -> int:
        result = []

        for s in tokens:
            if len(s)>1 or s[0].isdigit():
                result.append(int(s))
            else:
                x2 = result.pop();
                x1 = result.pop();
                
                if s[0]=='+': x1+=x2
                if s[0]=='-': x1-=x2
                if s[0]=='*': x1*=x2
                if s[0]=='/': x1/=x2

                result.append(x1)

        return result[-1]



if __name__ == "__main__":
    o = Solution()

    # INPUT
    tokens: list[str] = ["2","1","+","3","*"]

    # OUTPUT
    result = o.evalRPN(tokens)
    print(result)